infrastructure/smart-contracts/contracts/upgrades/StakingPool/StakingPool2.sol

```diff
@@ -0,0 +1,580 @@
+
+        uint256 userBalanceInKeyBucket = keyBucket.balanceOf(user);
+        uint256 userBalanceInEsXaiBucket = esXaiStakeBucket.balanceOf(user);
+
+        if (userBalanceInKeyBucket != 0) {
+            uint256 amountPerKey = amountForKeyBucket * 1_000_000 / keyBucket.totalSupply();
+			claimAmountFromKeys = amountPerKey * userBalanceInKeyBucket / 1_000_000;
+            claimAmount += claimAmountFromKeys;
+        }
+
+        if (userBalanceInEsXaiBucket != 0) {
+            uint256 amountPerStakedEsXai = amountForEsXaiBucket * 1_000_000 / esXaiStakeBucket.totalSupply();
+			claimAmountFromEsXai = amountPerStakedEsXai * userBalanceInEsXaiBucket / 1_000_000;
+            claimAmount += claimAmountFromEsXai;
+        }
+    }
+
+    function getUndistributedClaimAmount(
+        address user
+    ) external view returns (
+		uint256 claimAmountFromKeys,
+		uint256 claimAmountFromEsXai,
+		uint256 claimAmount,
+		uint256 ownerAmount
+	) {
+        return _getUndistributedClaimAmount(user);
+    }
+
+    function getPoolInfo()
+        external
+        view
+        returns (
+            PoolBaseInfo memory baseInfo,
+            string memory _name,
+            string memory _description,
+            string memory _logo,
+            string[] memory _socials,
+			uint32[] memory _pendingShares,
+			uint256 _ownerStakedKeys,
+			uint256 _ownerRequestedUnstakeKeyAmount,
+			uint256 _ownerLatestUnstakeRequestLockTime
+        )
+    {
+        baseInfo.poolAddress = address(this);
+        baseInfo.owner = poolOwner;
+        baseInfo.keyBucketTracker = address(keyBucket);
+        baseInfo.esXaiBucketTracker = address(esXaiStakeBucket);
+        baseInfo.keyCount = keyBucket.totalSupply();
+        baseInfo.totalStakedAmount = esXaiStakeBucket.totalSupply();
+        baseInfo.ownerShare = ownerShare;
+        baseInfo.keyBucketShare = keyBucketShare;
+        baseInfo.stakedBucketShare = stakedBucketShare;
+        baseInfo.updateSharesTimestamp = updateSharesTimestamp;
+
+        _name = name;
+        _description = description;
+        _logo = logo;
+        _socials = socials;
+
+        _pendingShares = new uint32[](3);
+        _pendingShares[0] = pendingShares[0];
+        _pendingShares[1] = pendingShares[1];
+        _pendingShares[2] = pendingShares[2];
+
+        _ownerStakedKeys = getCurrentKeyAmount(poolOwner);
+
+		_ownerRequestedUnstakeKeyAmount = userRequestedUnstakeKeyAmount[poolOwner];
+
+		if (_ownerStakedKeys == _ownerRequestedUnstakeKeyAmount && _ownerRequestedUnstakeKeyAmount > 0) {
+			_ownerLatestUnstakeRequestLockTime = unstakeRequests[poolOwner][unstakeRequests[poolOwner].length - 1].lockTime;
+		}
+    }
+
+    function getUserPoolData(
+        address user
+    )
+        external
+        view
+        returns (
+            uint256 userStakedEsXaiAmount,
+            uint256 userClaimAmount,
+            uint256[] memory userStakedKeyIds,
```

Comment:
Just a note on this one but there is an upper bound on the length of the return array, with tiny keys its possible we could hit this limit on a pool with too many keys.

Reply:
That is a good catch. This will now return keyAmounts. Suprisingly enough, this will not require too many changes throught the tests and clients.